shared/naturalcrit/codeEditor/helpers/widgets.js

```diff
@@ -1,38 +1,111 @@
 const React = require('react');
 const ReactDOM = require('react-dom');
-const { PATTERNS, FIELD_TYPE } = require('./widget-elements/constants');
+const { PATTERNS, FIELD_TYPE, HINT_TYPE, UNITS } = require('./widget-elements/constants');
 require('./widget-elements/hints/hints.jsx');
+const { Checkbox, Field } = require('./widget-elements');
+
+// See https://codemirror.net/5/addon/hint/css-hint.js for code reference
+const pseudoClasses = { 'active'           : 1, 'after'            : 1, 'before'           : 1, 'checked'          : 1, 'default'          : 1,
+	'disabled'         : 1, 'empty'            : 1, 'enabled'          : 1, 'first-child'      : 1, 'first-letter'     : 1,
+	'first-line'       : 1, 'first-of-type'    : 1, 'focus'            : 1, 'hover'            : 1, 'in-range'         : 1,
+	'indeterminate'    : 1, 'invalid'          : 1, 'lang'             : 1, 'last-child'       : 1, 'last-of-type'     : 1,
+	'link'             : 1, 'not'              : 1, 'nth-child'        : 1, 'nth-last-child'   : 1, 'nth-last-of-type' : 1,
+	'nth-of-type'      : 1, 'only-of-type'     : 1, 'only-child'       : 1, 'optional'         : 1, 'out-of-range'     : 1,
+	'placeholder'      : 1, 'read-only'        : 1, 'read-write'       : 1, 'required'         : 1, 'root'             : 1,
+	'selection'        : 1, 'target'           : 1, 'valid'            : 1, 'visited'          : 1
+};
+
+const genKey = (...args)=>args.join('-');
 
 module.exports = function(CodeMirror, widgets, cm, setHints) {
```

Comment:
cm is the codemirror instance, and CodeMirror is the whole library, right? Do we need to have CodeMirror as a parameter here? Could it be `require`'d as a dependency in this file instead?

Reply:
Y'know what yes, _but_ I need to pull its declaration out into its own JS file

Reply:
Done!